Adding adjustable time range to single stock chart

kamalq97 opened the pull request:
Changelog:
- In **urls.py** file, added new entry for historic single stock with a defined time range
- Defined custom `InvalidTimeRange` exception and used it in **stock_api.py**
- In **views.py** file, added optional `time_range` parameter with a default value of '1m' (one month)
- In **single_stock.html**, added tabs to show different time scales and the line chart color also changes (to red or green) depending on stock performance 
- Added time range unit tests to **test_stock_api.py**
- In **test_views.py**, one endpoint has been removed since it now triggers a different error response: 400 [bad request] for invalid time range instead of the previous 404 [not found]

kamalq97 committed:
Update urls.py

kamalq97 committed:
Update views.py

The `single_stock_historic` view has been modified

kamalq97 committed:
Update stock_service.py

Defined new exception `InvalidTimeRange`

kamalq97 committed:
Update stock_api.py

Raise `InvalidTimeRange` exception if time range does not appear on the allowed list

kamalq97 committed:
Update single_stock.html

Added tabs to show different time scales, line chart color also changes depending on stock performance

kamalq97 committed:
Update test_stock_api.py

Added tests for different time ranges

kamalq97 committed:
Update test_views.py

Removed one of the non-exisitant end points since it no longer uses the `exception.html` template and instead just gives the error code 400 (Bad request).

kamalq97 committed:
Merge branch 'master' into single_stock_time_range

kamalq97 committed:
Merge branch 'master' into single_stock_time_range

murradkh commented on myapp/templates/single_stock.html:
```diff
@@ -61,64 +61,81 @@
+        </div>
+        <ul class="nav nav-pills mb-3" id="pills-tab" role="tablist">
+            <li class="nav-item">
+                <a class="nav-link" data-toggle="pill" onclick="getHistoricData('1d')" role="tab" aria-controls="pills-home" aria-selected="true">1D</a>
```
you used AjAX here!  Good one!

murradkh commented on myapp/templates/single_stock.html:
```diff
@@ -61,64 +61,81 @@
-			});
-		});
-    });
+    $(document).ready(getHistoricData());
```
small note, the JavaScript getting bigger here in this file, preferred to be separated and to include into a new JSscript file in `static/js/` folder

murradkh commented on myapp/tests/test_stock_api.py:
```diff
@@ -10,35 +10,55 @@ def setUp(self):
+            if time_range.endswith('d'):
+                first_day = datetime.fromisoformat(response[0].get('date')).day
+                last_day = datetime.fromisoformat(response[-1].get('date')).day
+                self.assertEquals(last_day - first_day, int(time_range[0])-1)
```
Does this handle the issue you mentioned about Friday and Saturday having no stocks-reports?

murradkh reviewed (commented):
Good Job!
with respect to the client-side view, two small comments:
1)  the buttons don't look like a clickable option (to act as a button)
2) when we click on 1D / 1Y/ max, will look like a messy data representation, hard to read, and many unneeded data/points. what we need I think for example when clicking on "1Y", to see only points on each month, and if I need more information I pick smaller time-range options.
so if we can fix this would be great.

kamalq97 committed:
Merge branch 'master' into single_stock_time_range

kamalq97 committed:
Update stock_api.py

Added chart_interval variable to change the sampling rate

kamalq97 committed:
Update single_stock.html

Tab buttons are now more visible and '1D' chart displays correctly

kamalq97 committed:
Update test_stock_api.py

Taking the difference between two dates (instead of two ints) and checking to see if the data retured is from the last 3 days or newer

kamalq97 committed:
Update test_stock_api.py

Updating tests to account for weekends

kamalq97 committed:
Update test_stock_api.py

murradkh commented on myapp/templates/single_stock.html:
```diff
@@ -122,9 +138,13 @@
+                    options: 
+                        {
+                            responsive: true,
+                            spanGaps: true,
```
Nice! you fixed the null points In an easy way

murradkh commented on myapp/templates/single_stock.html:
```diff
@@ -61,64 +61,101 @@
+        });
+    }
+
+    function getGraphColors(p1, p2) {
+        if (p2 >= p1) {
+            return ['rgba(2, 144, 62, .2)', 'rgba(2, 144, 62, .5)']
+        }
+        else {
+            return ['rgba(220, 24, 24, .2)', 'rgba(220, 24, 24, .5)']
+        }
+    }
```
you can use random color to be returned whenever it's called, like this (copied):

> 
> function random_rgba() {
>     var o = Math.round, r = Math.random, s = 255;
>     return 'rgba(' + o(r()*s) + ',' + o(r()*s) + ',' + o(r()*s) + ',' + r().toFixed(1) + ')';
> }

I also would use this for stocks-comparison.

kamalq97 replied:
Okay, I will add this to my code

murradkh reviewed (approved):
working fine! good job  شغيل 👍 
I see that's you have many changes which overlap with mine in #33, many conflicts expected. so a lot of work will be. that's should not happen in general I think, we did a mistake by start working on features overlapping in code places.
next time we need to avoid this mistake.
for the current state, what I propose to do is that after you merge your PR, I will try to resolve the conflicts, and make sure my features are working properly, then you make sure your features are working properly (on the stocks-comparison branch)
what do you think?

kamalq97 committed:
Update single_stock.html

Now return random colour for the graph
